crates/bevy_ecs/src/schedule/run_criteria.rs

```diff
@@ -44,6 +41,17 @@ pub enum ShouldRun {
     NoAndCheckAgain,
 }
 
+impl ShouldRun {
+    pub fn once(mut ran: Local<bool>) -> ShouldRun {
```

Comment:
Ooh this is interesting. One more small consistency thing to discuss api on:

This uses "normal system" add-using-function-name syntax `add_system(foo.with_run_criteria(ShouldRun::once))`, which has the benefit of being exactly what you would do for a "non-closure system".

However if were were to add a theoretical `ShouldRun::n_times`, it would need to return a closure to store the value of `n`.

So we would have the inconsistent styles:

```rust
app.
  .add_system(foo.with_run_criteria(ShouldRun::once))
  .add_system(foo.with_run_criteria(ShouldRun::n_times(42))
```

So the question is: is this ok?  (this is not a loaded question ... i dont have strong opinions here). When using the "functions on structs" pattern for systems, what are our "rules"?
Should they be:
1. When using struct-functions, always use functions returning `impl Fn/FnMut` for consistency
2. When using struct-functions, prefer "normal system" style when possible and return `impl Fn/FnMut` only when required by the context.
3. When using struct-functions, prefer "normal system" style when possible, but prefer consistency across functions on the struct if there is a mismatch.
4. Use `normal_system` when possible. Only use struct-functions for closure-style systems.

(3) loses the organizational benefits of struct style (which would be very confusing), so thats the only option I'm actively against.

Reply:
5. Use const generics. `ShouldRun::n_times::<{42}>`.

I'd generally lean towards option 2. I think we need to do the same migration for `FixedTimestep`, but I hadn't done it yet because of stageless.

Reply:
My preference is 4: I think they're clearer in general (especially in the docs), and we should have strong reasons to use struct-functions.

3 is definitely bad.

Reply:
Arg sorry I inserted an extra list item, invalidating my reference to `(3)`. Sorry for the confusion (bad cart!). I meant to say:

> (4) loses the organizational benefits of struct style (which would be very confusing), so thats the only option I'm actively against.

Reply:
I think this would be actively bad UX:

```rust
app
  .add_system(foo.with_run_criteria(should_run_once)
  .add_system(foo.with_run_criteria(ShouldRun::n_times(42))
```

Reply:
> Use const generics. ShouldRun::n_times::<{42}>.

Good point. There are cases where this would significantly increase the amount of internal statefullness / Local hackery required (ex: a timer that ticks down from 42 rather than up from 0 would need to have an init phase enforced with locals). But it would have the benefit of not using function calls, so its worth considering. It feels like there might be cases where closures are still required though (ex: the input is set at runtime).

Reply:
This discussion shouldn't block this PR landing, right?

I think in a lot of cases, we can avoid having systems being externally configurable. At some point, we'll probably want something `Local`-like initialised from a const parameter. Not sure if that's possible at the moment.

My preference is option 3. Option 3 keeps the rules for struct scoped systems the same as module scoped systems.
It's worth noting that as far as I know, the `_system` suffix is idiomatic for functions which return system functions, which applies both when struct scoped and module scoped.

Reply:
> This discussion shouldn't block this PR landing, right?

Only in that it would be a breaking change to switch from the current style to a "return FnMut for everything on a struct" style. I do think "only use FnMut when you need it" is pretty consistent with what we do today (there is already precedent for the current pattern in bevy).

 So yeah Option 2 seems like a reasonable thing to default to. Lets roll with that and we can revisit later if we feel inclined.